### PR TITLE
Fix bad checks and race condition from actor_deaths and node_failures tests

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This ports bug fixes discovered while running the long-running ray ci tests.

The main fix is around worker handling. Previously, we assumed worker ports were unique. However, this is not the case. If a worker dies, a new worker can be created with its port in the meantime. This leads to a race condition where the new worker starts operating with the port of the old worker.

The workaround we use here is to (1) add the pid to the worker address to uniquely identify the worker, and (2) ignore stale requests to return the old dead worker in the node manager.

- [ ] Wait for https://github.com/ray-project/ray/pull/6412 to merge

### CODE REVIEW

File: src/ray/common/ray_config_def.h [hunk 1]
```diff
@@ -63,7 +63,7 @@ RAY_CONFIG(int64_t, worker_lease_timeout_milliseconds, 500)
 /// If set to a negative value, the heartbeats will not be sent.
 /// These are used to report active object IDs for garbage collection and
 /// to ensure that workers go down when the raylet dies unexpectedly.
-RAY_CONFIG(int64_t, worker_heartbeat_timeout_milliseconds, 1000)
```
Comment (edoakes) [inline src/ray/common/ray_config_def.h]: Actually we need these heartbeats even if we aren't sending any IDs to the raylet to ensure workers die when the raylet dies unexpectedly. See my change in https://github.com/ray-project/ray/pull/6412 instead.

File: src/ray/common/task/task.h [hunk 2]
```diff
@@ -9,7 +9,7 @@
 
 namespace ray {
 
-typedef std::function<void(const std::shared_ptr<void>, const std::string &, int,
+typedef std::function<void(const std::shared_ptr<void>, const std::string &, int, int,
```
Comment (edoakes) [inline src/ray/common/task/task.h]: This is getting complicated, can we add param names?

File: src/ray/core_worker/future_resolver.cc [hunk 3]
```diff
@@ -9,7 +9,7 @@ void FutureResolver::ResolveFutureAsync(const ObjectID &object_id, const TaskID
   auto it = owner_clients_.find(owner_id);
   if (it == owner_clients_.end()) {
     auto client = std::shared_ptr<rpc::CoreWorkerClientInterface>(
-        client_factory_({owner_address.ip_address(), owner_address.port()}));
+        client_factory_({owner_address.ip_address(), owner_address.port(), -1}));
```
Comment (edoakes) [inline src/ray/core_worker/future_resolver.cc]: comment for the -1

**Review by zhijunfu — CHANGES_REQUESTED** [review]
We already support multiple workers in the same process (for java), and `pid` itself cannot identify a worker any more. 

How about we use `worker_id` instead of `pid`?

File: src/ray/protobuf/node_manager.proto [hunk 4]
```diff
@@ -24,9 +24,11 @@ message WorkerLeaseReply {
 message ReturnWorkerRequest {
   // Port of the leased worker that we are now returning.
   int32 worker_port = 1;
+  // Pid of the leased worker we are now returning.
+  int32 worker_pid = 2;
```
Comment (zhijunfu) [inline src/ray/protobuf/node_manager.proto]: could we change to `worker_id` instead? since `pid` no longer identifies a worker with multi-threading.

**Review by zhijunfu — APPROVED** [review]
LGTM, thanks

**Pull request merged**